atc/config.go

```diff
@@ -296,6 +296,104 @@ func (types ResourceTypes) Without(name string) ResourceTypes {
 	return newTypes
 }
 
+func (types ResourceTypes) ImageForType(planID PlanID, resourceType string, stepTags Tags, skipInterval bool) TypeImage {
+	// Check if resource type is a custom type
+	parent, found := types.Lookup(resourceType)
+	if !found {
+		// If it is not a custom type, return back the image as a base type
+		return TypeImage{
+			BaseType: resourceType,
+		}
+	}
+
+	imageResource := ImageResource{
+		Name:   parent.Name,
+		Type:   parent.Type,
+		Source: parent.Source,
+		Params: parent.Params,
+		Tags:   parent.Tags,
+	}
+
+	getPlan, checkPlan := FetchImagePlan(planID, imageResource, types.Without(parent.Name), stepTags, skipInterval, parent.CheckEvery)
+	checkPlan.Check.ResourceType = resourceType
+
+	return TypeImage{
+		// Set the base type as the base type of its parent. The value of the base
+		// type will always be the base type at the bottom of the dependency chain.
+		//
+		// For example, if there is a resource that depends on a custom type that
+		// depends on a git base resource type, the BaseType value of the resource's
+		// TypeImage will be git.
+		BaseType: getPlan.Get.TypeImage.BaseType,
+
+		Privileged: parent.Privileged,
+
+		// GetPlan for fetching the custom type's image and CheckPlan
+		// for checking the version of the custom type.
+		GetPlan:   &getPlan,
+		CheckPlan: checkPlan,
+	}
+}
+
+func FetchImagePlan(planID PlanID, image ImageResource, resourceTypes ResourceTypes, stepTags Tags, skipInterval bool, checkEvery *CheckEvery) (Plan, *Plan) {
+	// If resource type is a custom type, recurse in order to resolve nested resource types
+	getPlanID := planID + "/image-get"
```

Comment:
If there are nested resource types, plan id could be `some-id/image-get/image-get/image-check/...`. To make it clearer, can we change to `some-id/image-get/<type name>/image-get/<other type name>/...`.

Not sure how plan id is used. If the change would lead to complex changes in UI, please just ignore this comment.

Reply:
The plan ID doesn't impact too much, so wouldn't be hard to change - that said, it's rarely surfaced to the user (the only place it appears is in the URL when you highlight lines on the build page)

Reply:
I'm not sure if it is necessary to add the type name into the plan_id though because it is just used as a unique identifier. I don't know who would be needing to decode the plan ids, as the plan id of steps in the build plan are just incremented numbers that don't have anything to do with the actual step itself. Did you have anything in mind for using the plan ids?